Add styling and additional logic to search for users

Alexander-Dubrawski opened the pull request:
This PR closes #127 and closes #98 .

In the navbar is now an icon that is redirecting the user to the search page. 

The styling is now looking as follows: 

Desktop:
 
<img width="50%" height="50%" alt="Screenshot 2021-01-29 at 22 31 56" src="https://user-images.githubusercontent.com/32880898/106329546-eb205500-6281-11eb-8d98-2015f111cd83.png">


Mobile: 

<img width="50%" height="50%" alt="Screenshot 2021-01-29 at 22 08 54" src="https://user-images.githubusercontent.com/32880898/106329571-f7a4ad80-6281-11eb-84fa-1e3abd77ac5c.png">

Paula-Kli commented on app/controllers/users_controller.rb:
```diff
@@ -44,12 +57,23 @@ def add_contact
+    @contacts = Set[]
+    if current_user && params[:search]
+      @user = User.find(current_user.id)
+      @users = search_record(params[:search], User, use_wildcards)
```
why are you not using the method from the Users model anymore? I don't think this is a good way to implement searching since you are using a method from the users helper but nevertheless have to tell which record you want to search in.
`User.search(params[:search]).where.not(id: current_user.id)` this line gave you all users which match the search parameter and which are not the current user

felixauringer replied:
I would strongly support this. The search logic built by Scrumbledore's Army is reviewed and well-tested. I do not see any reason to rebuild this logic. And I would also prefer the approach of the existing solution over your new approach.

Alexander-Dubrawski replied:
Hm... You are right. I don't know what happened there 🤔 . I will use the search logic from before.

Paula-Kli commented on app/controllers/users_controller.rb:
```diff
@@ -44,12 +57,23 @@ def add_contact
-    @users = User.search(params[:search]).where.not(id: current_user.id)
-    @users_to_add = @users.reject do |user|
-      current_user.sent_contact_request?(user)
+    use_wildcards = true
```
if you are just turning that to `true` to give that to the method, why are you not putting `true` into the method call?

felixauringer replied:
Because it is currently not necessary to switch wildcards off, I would completely omit this variable and always use wildcards.

Alexander-Dubrawski replied:
I replaced the search logic with the one from Scrumbledore's Army. So we are not using this variable anymore.

Paula-Kli commented on app/controllers/users_controller.rb:
```diff
@@ -44,12 +57,23 @@ def add_contact
+    @users = Set[]
+    @contacts = Set[]
+    if current_user && params[:search]
+      @user = User.find(current_user.id)
```
I think that variable could be named `@current_user` to be more precise

Alexander-Dubrawski replied:
Same as above.

Paula-Kli commented on app/views/layouts/_flash_messages.html.erb:
```diff
@@ -1,7 +1,10 @@
     <div class='alert alert-<%= level %>'>
-      <% if log.respond_to?(:each) %>
-        <% if log[:heading].nil? %>
+      <% if log.is_a? String %>
```
what does this change mean? why is this changed in the "change search" pr?

Alexander-Dubrawski replied:
It was just a quick fix to be able to continue implementing. But after the merge with the master, I replaced the code with the new code from there.

Paula-Kli commented on app/views/users/search.html.erb:
```diff
@@ -1,42 +1,51 @@
-            <%= label_tag(:search, I18n.t('user.search.for') ) %>
-            <%= text_field_tag :search, params[:search] %>
-            <%= submit_tag I18n.t('user.search.index') %>
+  <% @contacts.each do |user| %>
```
is there a reason, why the logic is changed? should I first see my contacts and later other people?

I would not call one contact just user but `contact` to know why they are handled differently

when #158  ist merged you can just call `<%= render partial: 'users/card', locals: { user: contact, mode: :overview } %>` here to get the expected card with the social accounts

Alexander-Dubrawski replied:
> is there a reason, why the logic is changed? should I first see my contacts and later other people?

Yes, that is the desired behavior.

Alexander-Dubrawski replied:
I would like to use the new design, especially since it is a pretty nice generic solution. But after some testing (Iphone 6/7/8), I am a little bit conflicted. The links don't seem to be placed correctly:  

<img width="372" alt="Screenshot 2021-02-03 at 20 59 19" src="https://user-images.githubusercontent.com/32880898/106802171-dddcdf00-6662-11eb-926e-94580bc22b7a.png">

I just changed the code to:

```rails
      <% @users.each do |user| %>
             <%= render partial: 'users/card', locals: { user: user, mode: :overview } %>
     <% end %>
```

Am I doing something wrong? 😞

Alexander-Dubrawski replied:
If you can't find a quick fix I could try to adjust it with some CSS.

Paula-Kli replied:
I dont think you are doing somethin wrong and for me with firefox and iPhone 6/7/8 it looks like this:
![grafik](https://user-images.githubusercontent.com/49535291/106805448-fbac4300-6666-11eb-9ec0-7c13357eb9c3.png)

Paula-Kli replied:
I changed the code, to the desired partial but I am on your branch

Alexander-Dubrawski replied:
@Paula-Kli there are a few problems with the cards that catch my eye:

* Why are we using tables to style the cards? Tables were used in the early days of CSS. Now usually flex or grids are used for that. With flex, you are able to place the content much nicer and evenly. For example, you can distribute the handles evenly. I think the problem that we saw is caused by using tables.
* Why are we using cards from Bootstrap. Bootstrap cards usually have a different layout. If we want to simply display a shadow and have round corners we can  use basic CSS

I could quickly adjust the code of your partials so that the problems are solved if it is ok for you 🙃 . We already have all the CSS.

Alexander-Dubrawski replied:
will be solved with #189 .

Paula-Kli commented on app/views/users/search.html.erb:
```diff
@@ -1,42 +1,51 @@
-            <% end %>
-          </tr>
+  <% @users.each do |user| %>
+    <div class=contact>
```
when #158  ist merged you can simply use `<%= render partial: 'users/card', locals: { user: user, mode: :add } %>` to get the expected card with a `+` button and a tooltip

felixauringer replied:
That is now merged :+1:

Alexander-Dubrawski replied:
Please have a look at the conversation above.

Paula-Kli commented on config/routes.rb:
```diff
@@ -5,6 +5,7 @@
   devise_for :users
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
   root to: 'home#dashboard'
+  get 'search', to: 'users#search'
```
why is this line added when the same is done in line 14?

Alexander-Dubrawski replied:
We need this route if we try to reload the search page. Not sure why it is not working with the one defined in line 14 though, to be honest.

Paula-Kli replied:
did you try using the path helper `search_users_path` as mentioned below it is not just `/search` but `users/search` and due to that the correct path is `search_users_path`

Alexander-Dubrawski replied:
Now it works, thanks 😄 .

Paula-Kli commented on app/controllers/users_controller.rb:
```diff
@@ -44,12 +57,23 @@ def add_contact
+    if current_user && params[:search]
+      @user = User.find(current_user.id)
+      @users = search_record(params[:search], User, use_wildcards)
+      @contacts = search_record(params[:search], @user.contacts, use_wildcards)
```
please use `current_user.sent_contact_request?` here because we not only want to send no second request to a user that is already our contact but also not to someone with a pending request

Alexander-Dubrawski replied:
Removed this search logic.

Paula-Kli commented:
should the empty search page maybe display all available users?

felixauringer commented on app/controllers/users_controller.rb:
```diff
@@ -44,12 +57,23 @@ def add_contact
     redirect_to root_path
   end
 
+  # rubocop:disable Metrics/AbcSize
```
I would prefer changing code not the config of rubocop. Why should we use it at all if we are turning it off when it detects opportunities for improvements?

Alexander-Dubrawski replied:
In my opinion, sometimes methods will be more complex than 10 lines of code. But I am programming quite a lot of c++ lately so maybe I just have a different subjective view on this matter. Since I replaced this method with the one of  Scrumbledore's Army I don't need to disable rubocop here.

felixauringer commented on app/assets/stylesheets/users.scss:
```diff
@@ -1,3 +1,86 @@
+    border-radius: 99999px;
+    background: #343B3F;
+    font-size: small;
+}
```
missing newline

felixauringer commented on app/controllers/users_controller.rb:
```diff
@@ -1,9 +1,18 @@
-
-  def show; end
+  include UsersHelper
+  before_action :authorize, except: %i[show index search]
```
Why do you change the authorization settings for this controller? I couldn't find it specified in the linked ticket and I think @LinqLover did really good work when introducing proper authentication.

Alexander-Dubrawski replied:
Not sure why I did it. I removed it.

felixauringer commented on app/helpers/users_helper.rb:
```diff
@@ -1,2 +1,25 @@
+require 'set'
+
 module UsersHelper
+  # rubocop:disable Metrics/MethodLength
```
You can't just disable Rubocop wheresoever it complains about something :sweat:

felixauringer commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -3,6 +3,7 @@
   <form class="form-inline">
     <i class="fa fa-bell" style="color: white; padding: 10px"></i>
-    <a href="<%= search_users_path %>"><i class="fa fa-search" style="color: white; padding: 10px"></i> </a>
+    <%= link_to (fa_icon 'search'), "/search", style:  "color: white; padding: 10px"%>
```
please use path helpers

felixauringer commented on app/views/users/search.html.erb:
```diff
@@ -1,42 +1,51 @@
+  <% @users.each do |user| %>
+    <div class=contact>
+      <label class="contact-info">
+        <div class="profile-image my-2 my-sm-0 ">
```
This is almost the same code as for contacts above

Alexander-Dubrawski replied:
I know 🙈 . Just wanted to wait for the implementation of  #158 . Otherwise, I would have implemented some genric partials.

felixauringer replied:
#158 has been merged now

felixauringer reviewed (changes requested):
Remove the `.DS_Store` file please. You could also add it to your global gitignore on your computer (not in the project's gitignore) to prevent committing this file in all git projects you are working on.

You should merge `dev` into this branch because it already contains a lot of changes that you tried to implement here. I would always wait for PRs to be merged instead of implementing them yourself. That duplicates work and increases the size of PRs so reviewing gets a lot harder.

So I think all the things mentioned in `Additional implementation` in the PR description should not be in this branch. For example the logout button: If you notice that it is missing, you could open a ticket, fix it with maybe 5 lines of code, write one test for it and the PR could be merged in one day. But if you include something like this in a PR as big as this one, it takes forever until users can log out again.

rheaSPK commented on app/views/layouts/_flash_messages.html.erb:
```diff
@@ -13,4 +13,4 @@
       <% end %>
     </div>
-<% end %>
+<% end %>
```
Please add new line

rheaSPK commented on app/views/layouts/_footer.html.erb:
```diff
@@ -1,29 +1,58 @@
+<style>
```
please put this a seperate css file like `mobile.scss`

rheaSPK reviewed (commented):
Please add the new card style, it's already merged and please fix CI-Errors, somethings not right with some of your code...

felixauringer commented on app/views/layouts/_footer.html.erb:
```diff
@@ -1,29 +1,58 @@
+      </li>
     </ul>
   </div>
 <% end %>
```
missing newline

felixauringer commented on app/views/layouts/_navbar.html.erb:
```diff
@@ -3,7 +3,8 @@
+    <%= link_to (fa_icon 'sign-out'), destroy_user_session_path, method: :delete, class: 'nav-link' %>
   </form>
 <%= render 'layouts/navbar_desktop_part' %>
 </nav>
```
missing newline

felixauringer commented on spec/features/user_search_spec.rb:
```diff
@@ -7,9 +7,16 @@
 
     before { visit search_users_path }
 
+    it 'is possible to filter by username' do
```
Why did you add this test? Exactly this behaviour is tested in the line 55 (before your changes) / 62 (after your changes).
If you want to check that the first name is visible, then I would add this expectation to all tests below

Alexander-Dubrawski replied:
I removed the test.

felixauringer commented on spec/views/users/search.html.erb_spec.rb:
```diff
@@ -2,15 +2,23 @@
   end
 
-  it 'renders a list of users before searching' do
+  it 'renders a list of contacts' do
```
What's your thought behind the changes in this test? You have renamed the variable in the test and added the existing test again with the renamed variable but that shouldn't change the behaviour of the test. I would therefore restore this file to its current state on dev

Alexander-Dubrawski replied:
We have *users* and *users_to_add* and we want to check that both groups are being rendered. That's why we have two test cases.

felixauringer replied:
That's a good idea, but that is not what the test does :sweat_smile:

felixauringer replied:
`assign(:users, contacts)` is like executing `@users = contacts` in the controller before rendering the view (    assign(:users, contacts))

Alexander-Dubrawski replied:
I just changed the code back to the dev status.

felixauringer commented on app/assets/stylesheets/users.scss:
```diff
@@ -1,3 +1,86 @@
 // Place all the styles related to the Users controller here.
 // They will automatically be included in application.css.
 // You can use Sass (SCSS) here: https://sass-lang.com/
+
```
Is this file still necessary with #158 merged?

Alexander-Dubrawski replied:
Yes, it is. We are using some styles for the search form.

felixauringer commented on app/controllers/users_controller.rb:
```diff
@@ -1,9 +1,14 @@
+require 'set'
```
why?

felixauringer commented on app/controllers/users_controller.rb:
```diff
@@ -1,9 +1,14 @@
   include SocialAccountsHelper
-  before_action :authorize
+  include UsersHelper
+  before_action except: %i[show index search]
```
What does this do? I would expect the symbol of a method name here that would be executed before the specified actions

Alexander-Dubrawski replied:
I added `before_action :authorize` again.

felixauringer commented on app/controllers/users_controller.rb:
```diff
@@ -1,9 +1,14 @@
 class UsersController < ApplicationController
   include SocialAccountsHelper
-  before_action :authorize
+  include UsersHelper
```
?
```suggestion
```

felixauringer commented on app/controllers/users_controller.rb:
```diff
@@ -1,9 +1,14 @@
   helper_method :generate_link, :supported_social_networks
 
-  def show; end
+  def show
```
This change should not be necessary if you would leave `before_action :authorize` this would not be necessary

Alexander-Dubrawski replied:
I changed it back

felixauringer commented on app/controllers/users_controller.rb:
```diff
@@ -31,6 +36,10 @@ def update_status
 
   def index
+    unless current_user
+      redirect_to(new_user_session_path)
```
This change should not be necessary if you would leave before_action :authorize this would not be necessary

Alexander-Dubrawski replied:
changed it back

frcroth commented on config/locales/en.yml:
```diff
@@ -71,6 +71,7 @@ en:
     search:
       index: 'Search'
       for: 'Search for:'
+      place_holder: 'Username, email, ...'
```
Use placeholder instead of place_holder

frcroth commented on app/views/layouts/_footer.html.erb:
```diff
@@ -1,29 +1,60 @@
-          <i class="fa fa-home"></i>
-        </a></li>
+         <label class="nav-bar-entry">
+          <!-- new path to this when dashboard is rewritten -->
```
This comment does not seem valid

frcroth commented on app/views/layouts/_footer.html.erb:
```diff
@@ -1,29 +1,60 @@
-        <a class="nav-link <%= active_link_to_class(chat_path, active: :exclusive)%>" href="<%= chat_path %>">
-          <i class="fa fa-comments"></i>
-        </a></li>
+        <!-- please update the path when reworked the chat -->
```
Remove

frcroth commented on app/views/layouts/_footer.html.erb:
```diff
@@ -1,29 +1,60 @@
-        </a></li>
+        <!-- please update the path when reworked the chat -->
+          <label class="nav-bar-entry">
+          <a class="nav-link footer-nav-link <%= active_link_to_class(root_path, active: :exclusive)%>" href="<%= chat_path %>">
```
Replace root_path with chat_path

frcroth commented on app/views/layouts/_footer.html.erb:
```diff
@@ -1,29 +1,60 @@
+          <label class="nav-bar-entry">
+          <a class="nav-link footer-nav-link <%= active_link_to_class(root_path, active: :exclusive)%>" href="<%= chat_path %>">
+            <i class="fa fa-comments"></i>
+          </a>
+        </label>
+      </li>
       <li class="nav-item">
-        <a class="nav-link <%= active_link_to_class(user_contacts_path(current_user), active: :exclusive)%>"
-           href="<%= user_contacts_path(current_user) %>">
-          <i class="fa fa-address-book"></i>
-        </a></li>
+        <label class="nav-bar-entry">
+          <a class="nav-link footer-nav-link <%= active_link_to_class(user_contacts_path(current_user), active: :exclusive)%>" 
+            href="<%= user_contacts_path(current_user) %>">
+            <i class="fa fa-address-book"></i>
+          </a>
+        </label>
+      </li>
       <li class="nav-item">
         <!-- please add the right path here when history view created -->
-        <a class="nav-link" href="">
-          <i class="fa fa-history"></i>
-        </a></li>
+        <label class="nav-bar-entry">
+          <a class="nav-link footer-nav-link" href="">
+            <i class="fa fa-history"></i>
+          </a>
+        </label>
+      </li>
       <li class="nav-item">
-        <a class="nav-link <%= active_link_to_class(edit_user_path(current_user), active: :exclusive)%>"
-           href="<%= edit_user_path(current_user) %>">
-          <i class="fa fa-cog"></i>
-        </a></li>
+        <label class="nav-bar-entry">
+          <a class="nav-link footer-nav-link <%= active_link_to_class(edit_user_path(current_user), active: :exclusive)%>" 
+            href="<%= edit_user_path(current_user) %>">
+            <i class="fa fa-cog"></i>
+          </a>
+        </label>
```
What is the reason for the changes in this file? It does not seem to be related to the topic of the PR. You could remove this here/ move it to another PR to reduce complexity

Alexander-Dubrawski replied:
Maybe the following screenshot describes it the best: 

<img width="1220" alt="Screenshot 2021-02-05 at 17 09 16" src="https://user-images.githubusercontent.com/32880898/107066455-4b614a80-67de-11eb-812a-1710c19e730d.png">

The link of the setting page is covering the whole page.

Alexander-Dubrawski replied:
To solve this problem I simply wrap the link inside a label.

frcroth replied:
Sorry I don't have access to a PC right now. Can you reproduce this @felixauringer?

Alexander-Dubrawski replied:
@frcroth is it okay if I leave it in this PR? I know it's not the best solution, but at the moment I prefer the fastest one.

felixauringer replied:
![image](https://user-images.githubusercontent.com/48409110/107075179-0cd18d00-67ea-11eb-9377-ca183220cd61.png)

I'm on your branch @Alexander-Dubrawski but I resetted `app/views/layouts/_footer.html.erb` to its state on the `dev` branch and everything seems to be fine without your changes in this file. I'm not quite sure what the problem is :sweat_smile:

Alexander-Dubrawski replied:
That's confusing. I have the same problem on dev.  But I will double-check.

Alexander-Dubrawski replied:
Well..., somehow it got magically fixed. I can't reproduce it anymore either. Maybe one of the last merges fixed it. Its getting hard to keep track of.

felixauringer commented on spec/features/user_search_spec.rb:
```diff
@@ -7,6 +7,7 @@
 
   describe 'without filter' do
     let!(:users) { FactoryBot.create_list(:user, 3) }
+    let(:second_user) { FactoryBot.create :user }
```
Is this user still needed? It is the only change in this file, so I assume that it can be omitted?
```suggestion
```

felixauringer commented on spec/views/users/search.html.erb_spec.rb:
```diff
@@ -2,15 +2,23 @@
 RSpec.describe 'users/search', type: :view do
   let(:users) { FactoryBot.create_list(:user, 3) }
+  let(:contacts) { FactoryBot.create_list(:user, 3) }
+  let(:user1) { FactoryBot.create :user }
```
```suggestion
  let(:user) { FactoryBot.create :user }
```

felixauringer commented on spec/views/users/search.html.erb_spec.rb:
```diff
@@ -2,15 +2,23 @@
 
   before do
-    assign(:users, users)
+    assign(:users, contacts)
```
```suggestion
    assign(:users, contacts + users)
```

felixauringer commented on spec/views/users/search.html.erb_spec.rb:
```diff
@@ -2,15 +2,23 @@
+    assign(:users, contacts)
     assign(:users_to_add, users)
-    sign_in users.first
+    sign_in user1
```
```suggestion
    sign_in user
```

felixauringer reviewed (commented):
I added some suggestions that assign the instance variables in the view as the users controller would do

felixauringer commented:
The code looks great now! The PR has also a size that is very good reviewable. Just one question:
```ruby
def search
    @users = User.search(params[:search]).where.not(id: current_user.id)
    @users_to_add = @users.reject do |user|
      current_user.sent_contact_request?(user)
    end
  end
```
It looks like `users_to_add` is a subset of `users`. Did you try out whether users that aren't contacts of (or requested by) the current user are displayed twice?
If yes, I would suggest something like the following
```ruby
def search
    other_users = User.search(params[:search]).where.not(id: current_user.id)
    @users_to_add = other_users.reject do |user|
      current_user.sent_contact_request?(user)
    end
    @users = other_users.select do |user|
      current_user.sent_contact_request?(user)
    end
  end
```

Alexander-Dubrawski commented:
> The code looks great now! The PR has also a size that is very good reviewable. 

Sorry for this chaotic PR of mine 🙈 . I really don't know why I implemented stuff that was already existing. But thanks for all your feedback and patience. 
  

> It looks like `users_to_add` is a subset of `users`. Did you try out whether users that aren't contacts of (or requested by) the current user are displayed twice?

You are absolutely right. Good eye. I adjusted the code.

felixauringer reviewed (approved):
Nice to finally finish this :tada: I'm sorry for having so much comments but thank you for implementing all of it :)

frcroth commented:
![image](https://user-images.githubusercontent.com/6863832/107120080-c2085180-688b-11eb-95f9-5871de4aabf6.png)
This is what /users/search looks for me on the deployed version

frcroth commented:
@Alexander-Dubrawski

Alexander-Dubrawski commented:
> This is what /users/search looks for me on the deployed version

I can not reproduce the error. What exactly did you do? (For example: login, click the search icon, ...)

felixauringer commented:
> ![image](https://user-images.githubusercontent.com/6863832/107120080-c2085180-688b-11eb-95f9-5871de4aabf6.png)
> This is what /users/search looks for me on the deployed version

Same for me, I will try to investigate the reason

Alexander-Dubrawski commented:
> Same for me, I will try to investigate the reason

@felixauringer thanks that would be really nice. If you can not find the error, please ping me and I can also have a look this evening.

felixauringer commented:
Works locally for me, so I will have a look at the heroku logs

felixauringer commented:
The relevant part is

```text
ActionView::Template::Error (Can't resolve image into URL: to_model delegated to attachment, but attachment is nil):

2021-02-06T15:00:46.081546+00:00 app[web.1]:      3:     <div class="container">

2021-02-06T15:00:46.081546+00:00 app[web.1]:      4:       <div class="row">

2021-02-06T15:00:46.081547+00:00 app[web.1]:      5:         <div class="col-xs">

2021-02-06T15:00:46.081548+00:00 app[web.1]:      6:           <%= image_tag user.avatar, height: 65, width: 65, class: "rounded-circle" %>

2021-02-06T15:00:46.081548+00:00 app[web.1]:      7:         </div>

2021-02-06T15:00:46.081554+00:00 app[web.1]:      8:         <a class="col-8" href=<%= user_path(user) %>>

2021-02-06T15:00:46.081555+00:00 app[web.1]:      9:           <h5 class="card-title"><%= user.display_name %></h5>

2021-02-06T15:00:46.081556+00:00 app[web.1]:   

2021-02-06T15:00:46.081556+00:00 app[web.1]:  app/views/users/_card.html.erb:6

2021-02-06T15:00:46.081556+00:00 app[web.1]:  app/views/users/search.html.erb:14

2021-02-06T15:00:46.081557+00:00 app[web.1]:  app/views/users/search.html.erb:13:in `each'

2021-02-06T15:00:46.081557+00:00 app[web.1]:  app/views/users/search.html.erb:13
```

felixauringer commented:
So again problems with the avatars. This PR did not touch `_card.html.erb` so this problem was likely there before merging this one

felixauringer commented:
That's executed on production data:

```ruby
User.all.select { |user| user.avatar.nil? }.length
D, [2021-02-06T15:08:13.259421 #4] DEBUG -- :   User Load (1.8ms)  SELECT "users".* FROM "users"
=> 0
```
So the migration worked as expected

felixauringer commented:
I added an issue for this: #191
